src/librustc_metadata/dynamic_lib.rs

```diff
@@ -51,51 +51,97 @@ mod tests;
 
 #[cfg(unix)]
 mod dl {
-    use std::ffi::{CStr, CString, OsStr};
+    use std::ffi::{CString, OsStr};
     use std::os::unix::prelude::*;
-    use std::ptr;
-    use std::str;
 
-    pub(super) fn open(filename: &OsStr) -> Result<*mut u8, String> {
-        check_for_errors_in(|| unsafe {
-            let s = CString::new(filename.as_bytes()).unwrap();
-            libc::dlopen(s.as_ptr(), libc::RTLD_LAZY) as *mut u8
-        })
-    }
+    // `dlerror` is process global, so we can only allow a single thread at a
```

Comment:
Nit: `dlerror` It is MT-safe (thread local) on most modern POSIX implementations. Exceptions include freebsd and dragonflybsd as per investigation I did early this year, and apparently illumos as I’ve just learnt.

POSIX in particular allows for `dlerror` to be MT-safe (since a comparatively recent standard revision), but does not require it.